File: src/platform/audio_events.zig
const std = @import("std");

pub const AudioPermissionState = enum {
    unknown,
    prompt,
    granted,
    denied,
};

pub const AudioEventSegmentKind = enum {
    user_speech,
    assistant_speech,
    silence,
    sound_effect,
    background_audio,
    tool_call,
    interruption,
};

pub const AudioToolCallStatus = enum {
    pending,
    running,
    completed,
    failed,
};

pub const AudioInterruptionReason = enum {
    user_barge_in,
    playback_cancelled,
    timeout,
    system_cancelled,
};

pub const AudioPermissionEvent = struct {
    sequence_id: ?[]const u8 = null,
    state: AudioPermissionState,
};

pub const AudioInputLevelEvent = struct {
    sequence_id: []const u8,
    level_normalized: f32,
    peak_normalized: ?f32 = null,
    at_ms: u64,
};

pub const AudioSegmentStartedEvent = struct {
    sequence_id: []const u8,
    segment_id: []const u8,
    kind: AudioEventSegmentKind,
    start_ms: u64,
};

pub const TranscriptDeltaEvent = struct {
    sequence_id: []const u8,
    segment_id: []const u8,
    text: []const u8,
    is_final: bool = false,
};

pub const AudioDeltaEvent = struct {
    sequence_id: []const u8,
    segment_id: []const u8,
    audio_base64: ?[]const u8 = null,
    byte_len: ?u32 = null,
};

pub const ToolCallEvent = struct {
    sequence_id: []const u8,
    segment_id: []const u8,
    tool_name: []const u8,
    status: AudioToolCallStatus = .pending,
    latency_ms: ?u64 = null,
};

pub const InterruptionEvent = struct {
    sequence_id: []const u8,
    target_segment_id: []const u8,
    at_ms: u64,
    reason: AudioInterruptionReason,
};

pub const AudioSegmentCompletedEvent = struct {
    sequence_id: []const u8,
    segment_id: []const u8,
    end_ms: u64,
};

pub const SequenceCompletedEvent = struct {
    sequence_id: []const u8,
    completed_at_ms: u64,
};

pub const AudioErrorEvent = struct {
    sequence_id: ?[]const u8 = null,
    segment_id: ?[]const u8 = null,
    code: []const u8,
    message: []const u8,
    at_ms: u64,
};

pub const AudioEvent = union(enum) {
    permission_changed: AudioPermissionEvent,
    input_level: AudioInputLevelEvent,
    segment_started: AudioSegmentStartedEvent,
    transcript_delta: TranscriptDeltaEvent,
    audio_delta: AudioDeltaEvent,
    tool_call: ToolCallEvent,
    interruption: InterruptionEvent,
    segment_completed: AudioSegmentCompletedEvent,
    sequence_completed: SequenceCompletedEvent,
    error_: AudioErrorEvent,
};

test "audio event union carries normalized payloads" {
    const event: AudioEvent = .{
        .segment_started = .{
            .sequence_id = "seq_001",
            .segment_id = "a1",
            .kind = .assistant_speech,
            .start_ms = 1_200,
        },
    };

    switch (event) {
        .segment_started => |payload| {
            try std.testing.expectEqualStrings("seq_001", payload.sequence_id);
            try std.testing.expectEqual(.assistant_speech, payload.kind);
        },
        else => return error.UnexpectedEventTag,
    }
}

File: examples/ternary-bonsai/bridge/src/audio-events.mjs
/**
 * Normalized AudioEvent helpers matching `src/platform/audio_events.zig`.
 * The Zig runtime owns sequence state; this host only produces the seam payloads.
 */

export function nowMs(originMs) {
  return Date.now() - originMs;
}

export function makeSequenceId() {
  return `seq_${Date.now().toString(36)}`;
}

export function permissionChanged(sequenceId, state = "granted") {
  return {
    type: "permission_changed",
    sequence_id: sequenceId,
    state,
  };
}

export function segmentStarted(sequenceId, segmentId, kind, startMs) {
  return {
    type: "segment_started",
    sequence_id: sequenceId,
    segment_id: segmentId,
    kind,
    start_ms: startMs,
  };
}

export function transcriptDelta(sequenceId, segmentId, text, isFinal = false) {
  return {
    type: "transcript_delta",
    sequence_id: sequenceId,
    segment_id: segmentId,
    text,
    is_final: isFinal,
  };
}

export function segmentCompleted(sequenceId, segmentId, endMs) {
  return {
    type: "segment_completed",
    sequence_id: sequenceId,
    segment_id: segmentId,
    end_ms: endMs,
  };
}

export function toolCall(sequenceId, segmentId, toolName, status, latencyMs = null) {
  return {
    type: "tool_call",
    sequence_id: sequenceId,
    segment_id: segmentId,
    tool_name: toolName,
    status,
    latency_ms: latencyMs,
  };
}

export function sequenceCompleted(sequenceId, completedAtMs) {
  return {
    type: "sequence_completed",
    sequence_id: sequenceId,
    completed_at_ms: completedAtMs,
  };
}

export function errorEvent(code, message, atMs, sequenceId = null, segmentId = null) {
  return {
    type: "error_",
    sequence_id: sequenceId,
    segment_id: segmentId,
    code,
    message,
    at_ms: atMs,
  };
}

/** Collapse SSE token chunks into coarse transcript deltas for the panel. */
export function chunkText(buffer, next, minChars = 24) {
  const combined = buffer + next;
  if (combined.length < minChars) {
    return { emit: null, rest: combined };
  }
  return { emit: combined, rest: "" };
}
